PublicMapping/districtbuilder
# Allow saving with undo/redo

## Overview
This change makes undo/redo work with effects so that we can go backwards and forwards to previously saved map states. 

### Checklist

### Demo
![db_undo_redo_saving](https://user-images.githubusercontent.com/2926237/98174189-7a57df80-1ec2-11eb-9f74-7a75e452a17b.gif)

### Notes
This was really quite hard to get working... I've added longer comments around the trickier parts of this to try to make things more clear.

The big change in terms of what state we represent is to have both a `state` and optionally `effect` represented to undo/redo saves. We also now need to keep track of `districtsDefinition` in our `UndoableState` to allow for undoing/redoing of saves. Most of the rest of the code is state/effect management in the reducers as you might suspect.

There were several non-obvious things that tripped me up along the way to getting this working, such as how we were always clearing the selected geounits after a save (this worked fine until we wanted to not do that for undo/redo and thus had to be wired up differently for the undo/redo save case). There were several such small-ish changes needed to the reducers to make things shake out correctly, such as having a separate action just for updating `saving` state. Most of these changes are in or around `updateDistrictsDefinition`, so taking a look at that plus the comments I've added there should hopefully make things clear.

I'm a little concerned about the complexity of managing undo/redo state here but I'm not sure how it can really be avoided. Having the notion of a `Cmd` in redux-loop I found helpful as a description of an effect that can be kept lying around in the queue until we really need to apply it. One of the things I did to try to make this less brittle is to modify `pushState` and `replaceState` to work with essentially patches (single discrete changes) and not wholesale replacements (PATCH vs PUT in REST-speak). It seems that in all cases we just want to change one thing at a time so this should be relatively ergonomic (before this change, I was having trouble updating the whole piece of state where I had only meant to update a part of it) 

There is a refactor in this (see 4e307e3) which is to move the map management around undo/redo-ing geounit selection into the map itself. IMHO it's preferable for the reducer to just be responsible for state updates (and effects that relate to state) and then the map should sort itself out so as to look right in response to those state changes.

In closing, I'm glad that this seems to work but would welcome further abstraction/guard rails to make things more ergonomic/less brittle. I'm also open to having horribly over-complicated this so if there are any suggestions to simplify things I'm open to them.

## Testing Instructions
* Select a bunch of geounits
* Save them to a district
* Undo one step and they should be unsaved (the selected geounits should be back assigned to whatever district they were before) and the geounits should still be selected
* Redo and the geounits should be re-saved to the district you originally assigned them to
* Undo all the way back and your changes should be completely undone
* Redo all the way back and your changes should be completely redone
* Undo partially, select some other geounit, and then save again and it should work as expected

Closes #405
